### PR TITLE
Update MP REST Client to 4.0

### PR DESCRIPTION
**Pull request by radcortez**
<summary>MicroProfile REST Client Release notes:</summary>
<br>
<blockquote>
    <h2>4.0</h2>
    <ul>
        <li>New overloaded `baseUri(String uri)` method so users don’t have to call URI.create()</li>
        <li>New `header(String name, Object value)` method for adding headers to Client instances</li>
        <li>Clarify specification CDI support (`@RestClient` qualifier is not optional)</li>
        <li>Add multi-part processing example to specification.</li>
        <li>Clarify the specification that when a client injected as a CDI bean falls out of scope, the client is closed</li>
    </ul>
</blockquote>

[MicroProfile Rest Client Spec PDF](https://download.eclipse.org/microprofile/microprofile-rest-client-4.0/microprofile-rest-client-spec-4.0.pdf)
[MicroProfile Rest Client Spec HTML](http://download.eclipse.org/microprofile/microprofile-rest-client-4.0/microprofile-rest-client-spec-4.0.html)
[MicroProfile Rest Client Spec Javadocs](http://download.eclipse.org/microprofile/microprofile-rest-client-4.0/apidocs/)

### CODE REVIEW

File: extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/DefaultClientHeadersRequestFilter.java
Comment (gsmet) [inline extensions/resteasy-reactive/rest-client/runtime/src/main/java/io/quarkus/rest/client/reactive/runtime/DefaultClientHeadersRequestFilter.java:21]: Not sure about the use of the lambda here given it could be a hot path.

File: tcks/microprofile-rest-client-reactive/pom.xml
Comment (gsmet) [inline tcks/microprofile-rest-client-reactive/pom.xml:85]: Is it something we need to consider implementing? Question for both you and @geoand .

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java
Comment (gsmet) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java:388]:
```suggestion
                            // HttpClosedException / DecoderException are runtime exceptions. If we complete with that
```

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java
Comment (gsmet) [inline independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java:386]: Are you sure we should catch all `RuntimeException` here? The comment makes is seem like it should only be the two exceptions mentioned?

But maybe I'm being too cautious, I will defer to @geoand on that.

**Review by gsmet — COMMENTED** [review]
Thanks! It looks mostly good to me except for some minor suggestions.

I would prefer for @geoand to have a look though.

File: tcks/microprofile-rest-client-reactive/pom.xml
Comment (geoand) [reply tcks/microprofile-rest-client-reactive/pom.xml:85]: What does this do @radcortez ?

File: extensions/resteasy-classic/resteasy-client/runtime/src/main/java/io/quarkus/restclient/runtime/QuarkusRestClientBuilder.java
Comment (geoand) [inline extensions/resteasy-classic/resteasy-client/runtime/src/main/java/io/quarkus/restclient/runtime/QuarkusRestClientBuilder.java:169]: So this yet another way for users to be able to set headers?

File: independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java
Comment (radcortez) [reply independent-projects/resteasy-reactive/client/runtime/src/main/java/org/jboss/resteasy/reactive/client/handlers/ClientSendRequestHandler.java]: Yes, I was thinking about that, but I also didn't want to end up with a huge list of exceptions there.

File: extensions/resteasy-classic/resteasy-client/runtime/src/main/java/io/quarkus/restclient/runtime/QuarkusRestClientBuilder.java
Comment (radcortez) [reply extensions/resteasy-classic/resteasy-client/runtime/src/main/java/io/quarkus/restclient/runtime/QuarkusRestClientBuilder.java:169]: Yes, they added that in the builder API.

File: tcks/microprofile-rest-client-reactive/pom.xml
Comment (radcortez) [reply tcks/microprofile-rest-client-reactive/pom.xml:85]: This tests the ` EntityPart.Builder`, which we don't have an implementation for in Quarkus REST:
https://github.com/quarkusio/quarkus/blob/07b3f949d3df35f796dc3124e1e3f72249fe1502/independent-projects/resteasy-reactive/common/runtime/src/main/java/org/jboss/resteasy/reactive/common/jaxrs/RuntimeDelegateImpl.java#L145-L149

I'm not too concerned, especially because we can't use Quarkus REST to certify at this time. We could provide a proper implementation, but that can be done as a separate PR.

File: extensions/resteasy-classic/resteasy-client/runtime/pom.xml
Comment (geoand) [inline extensions/resteasy-classic/resteasy-client/runtime/pom.xml:64]: Not that I mind, but I would like to know why this is now necessary

File: extensions/resteasy-classic/resteasy-client/runtime/src/main/java/io/quarkus/restclient/runtime/QuarkusRestClientBuilder.java
Comment (geoand) [inline extensions/resteasy-classic/resteasy-client/runtime/src/main/java/io/quarkus/restclient/runtime/QuarkusRestClientBuilder.java:838]: I'm definitely no fan of this, but we can let it go in :)

File: tcks/microprofile-rest-client-reactive/pom.xml
Comment (geoand) [reply tcks/microprofile-rest-client-reactive/pom.xml:85]: Sure yeah, thanks

**Review by geoand — APPROVED** [review]
Thanks!!

File: extensions/resteasy-classic/resteasy-client/runtime/pom.xml
Comment (gsmet) [reply extensions/resteasy-classic/resteasy-client/runtime/pom.xml:64]: Also if it's necessary, we might have to depend on the `quarkus-resteasy-multipart` extension instead.

Comment (radcortez) [reply extensions/resteasy-classic/resteasy-client/runtime/pom.xml:64]: This is the dependency from RESTEasy that provides the `EntityPart.Builder` implementation. Note that for the REST Client Classic, I've tried to keep it as the original implementation: https://github.com/resteasy/resteasy-microprofile/commit/bb8b433fc28d0be2560d21d1204eec07603b9bf7

File: extensions/resteasy-classic/resteasy-client/runtime/src/main/java/io/quarkus/restclient/runtime/QuarkusRestClientBuilder.java
Comment (radcortez) [reply extensions/resteasy-classic/resteasy-client/runtime/src/main/java/io/quarkus/restclient/runtime/QuarkusRestClientBuilder.java:838]: Yes, this is a copy from the original RESTEasy Builder code:
https://github.com/resteasy/resteasy-microprofile/commit/cb58ceb333396875032aeeb58d3723569edcbb6b

Initially, we used the one provided by RESTEasy, but we made a copy to remove references to their CDI integration and also to remove the CDI dependency. They diverged a bit, so I've tried to reconcile them again for consistency.

**Comment by rsvoboda** [comment]
No description, noteworthy label, or indication if there are implications on end-user applications ...
@radcortez would you please cover that?

**Comment by michalvavrik** [comment]
I have opened issue here https://github.com/quarkusio/quarkus/issues/44067 because it cost me time to find this. Please document this. There is no PR description, there is migration guide, there is no documentation change. Thank you

**Comment by radcortez** [comment]
> No description, noteworthy label, or indication if there are implications on end-user applications ... @radcortez would you please cover that?

The reported issue in https://github.com/quarkusio/quarkus/issues/44067 is not caused by this change, which is a very simple one with zero impact on the end user. 

It is caused by https://github.com/quarkusio/quarkus/pull/43345, which completely describes the changes, including the breaking changes.

**Comment by radcortez** [comment]
No worries, all good. I usually include a Breaking Changes description in the PR, but only in the Migration Guide after the PR is merged because we can't be sure which version the PR will merge. https://github.com/quarkusio/quarkus/pull/43345 was merged recently, so it was still on my TODO list to copy the breaking changes to the Migration Guide.

**Comment by rsvoboda** [comment]
About stuff mentioned in https://github.com/quarkusio/quarkus/pull/43959#issuecomment-2434591612, noteworthy label should be applied, wdyt? 
And some description / link would be handy too, this is major version change.

**Comment by radcortez** [comment]
Updated!

**Pull request merged**
